core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala

```diff
@@ -0,0 +1,148 @@
+  case object Time extends PeriodMarkerName("time", classOf[TimeDownsamplePeriodMarker])
+  case object Counter extends PeriodMarkerName("counter", classOf[CounterDownsamplePeriodMarker])
+}
+
+/**
+  * Implementations of the trait identify row numbers that
+  * demark the boundaries of downsampling periods
+  */
+trait DownsamplePeriodMarker {
+
+  def name: PeriodMarkerName
+
+  /**
+    * Ids of Data Columns the marker works on.
+    * The column id values are fed in via downsampling configuration of the dataset
+    */
+  def inputColId: Int
+
+  /**
+    * Places row numbers for the given chunkset which marks the
+    * periods to downsample into the result buffer param
+    */
+  def getPeriods(part: ReadablePartition,
+                 chunkset: ChunkSetInfoReader,
+                 resMillis: Long,
+                 startRow: Int,
+                 endRow: Int): Buffer[Int]
+}
+
+/**
+  * Time based downsample period marker, which returns row numbers of the last sample for each downsample period
+  * @param inputColId requires the timestamp column 0
+  */
+class TimeDownsamplePeriodMarker(val inputColId: Int) extends DownsamplePeriodMarker {
+  require(inputColId == DataSchema.timestampColID)
+  override def getPeriods(part: ReadablePartition,
+                          chunkset: ChunkSetInfoReader,
+                          resMillis: Long,
+                          startRow: Int,
+                          endRow: Int): Buffer[Int] = {
+    val tsAcc = chunkset.vectorAccessor(DataSchema.timestampColID)
+    val tsPtr = chunkset.vectorAddress(DataSchema.timestampColID)
+    val tsReader = part.chunkReader(DataSchema.timestampColID, tsAcc, tsPtr).asLongReader
+
+    val startTime = tsReader.apply(tsAcc, tsPtr, startRow)
+    val endTime = tsReader.apply(tsAcc, tsPtr, endRow)
+
+    val result = debox.Buffer.empty[Int]
+    // A sample exactly for 5pm downsampled 5-minutely should fall in the period 4:55:00:001pm to 5:00:00:000pm.
+    // Hence subtract - 1 below from chunk startTime to find the first downsample period.
+    // + 1 is needed since the startTime is inclusive. We don't want pStart to be 4:55:00:000;
+    // instead we want 4:55:00:001
+    var pStart = ((startTime - 1) / resMillis) * resMillis + 1
+    var pEnd = pStart + resMillis // end is inclusive
+    // for each downsample period
+    while (pStart <= endTime) {
+      // fix the boundary row numbers for the downsample period by looking up the timestamp column
+      val endRowNum = Math.min(tsReader.ceilingIndex(tsAcc, tsPtr, pEnd), chunkset.numRows - 1)
+      result += endRowNum
+      pStart += resMillis
+      pEnd += resMillis
+    }
+    result
+  }
+  override def name: PeriodMarkerName = PeriodMarkerName.Time
+}
+
+/**
+  * Returns union of the following:
+  * (a) the results from TimeDownsamplePeriodMarker.
+  * (b) row number when counter drops
+  * (c) last row number before counter drops
+  * @param inputColId requires the counter column id
```

Comment:
Perhaps add a line or two explaining why we need to add the rows before/after drops